### ISSUE TITLE
rendered Lexy templates mess up indentation

### ISSUE THREAD

**Issue by raffaelj** [agentejo/cockpit#1244]
I like the Lexy template parser, but I don't like the messed up html indentation.

**fix:**
https://github.com/raffaelj/cockpit/commit/9ad961ebe5ca4b3472ffc396d5fde93f518f29d7

Find php tags after line breaks with following white space and replace them without the white space. Multiple line breaks are converted to a single linebreak. The regex fails for php tags with white space at the first line of the file.

```php
preg_replace('/([\r\n])(\s*)\<\?php/', '$1<?php', $value);
```

**example template:**

```php
    <p>one line test @if(1==1) one @elseif(2==2) two @else three @endif</p>
    <p>nested multi line test:</p>
    {% $arr = [1,2,3]; %}
    <ul>
    @if(count($arr))

      @foreach($arr as $i)
        <li>
        @if($i==1)
        one
        @elseif($i==2)
        two
        @else
        three
            <ul>
              @foreach([1,2] as $v)
                <li>{{ $i*$v }}</li>
              @endforeach
            </ul>
        @endif
        </li>
      @endforeach

    @endif
    </ul>
```

## current output

**rendered php:**

```php
    <p>one line test <?php if (1==1) { ?> one <?php } elseif(2==2) { ?> two <?php } else { ?> three <?php } ?></p>
    <p>nested multi line test:</p>
    <?php $arr = [1,2,3]; ?>
    <ul>
    <?php if (count($arr)) { ?>

      <?php foreach ($arr as $i) { ?>
        <li>
        <?php if ($i==1) { ?>
        one
        <?php } elseif($i==2) { ?>
        two
        <?php } else { ?>
        three
            <ul>
              <?php foreach ([1,2] as $v) { ?>
                <li><?php echo  $i*$v ; ?></li>
              <?php } ?>
            </ul>
        <?php } ?>
        </li>
      <?php } ?>

    <?php } ?>
    </ul>
```

**ugly html output:**

```html
    <p>one line test  one </p>
    <p>nested multi line test:</p>
        <ul>
    
              <li>
                one
                </li>
              <li>
                two
                </li>
              <li>
                three
            <ul>
                              <li>3</li>
                              <li>6</li>
                          </ul>
                </li>
      
        </ul>
```

## output after my fix

**rendered php:**

```php
    <p>one line test <?php if (1==1) { ?> one <?php } elseif(2==2) { ?> two <?php } else { ?> three <?php } ?></p>
    <p>nested multi line test:</p>
<?php $arr = [1,2,3]; ?>
    <ul>
<?php if (count($arr)) { ?>
<?php foreach ($arr as $i) { ?>
        <li>
<?php if ($i==1) { ?>
        one
<?php } elseif($i==2) { ?>
        two
<?php } else { ?>
        three
            <ul>
<?php foreach ([1,2] as $v) { ?>
                <li><?php echo  $i*$v ; ?></li>
<?php } ?>
            </ul>
<?php } ?>
        </li>
<?php } ?>
<?php } ?>
    </ul>
```

**pretty html output:**

```html
    <p>one line test  one </p>
    <p>nested multi line test:</p>
    <ul>
        <li>
        one
        </li>
        <li>
        two
        </li>
        <li>
        three
            <ul>
                <li>3</li>
                <li>6</li>
            </ul>
        </li>
    </ul>
```

**Comment by raffaelj** [comment]
I moved the discussion from the closed PR #1245 to this issue:

> Thank you for the contribution Raffael!
But because this is more an opinionated cosmetical improvement, I would like to politely decline this PR.

> You can still achieve this btw by adding the following snippet to an addon:

```php
cockpit()->renderer->extend(function($content) {
    return preg_replace('/([\r\n])(\s*)\<\?php/', '$1<?php', $content);
});
```

> I understand, that you decline it. The problem is, that I can't extend it that way. The user extensions are called before the internal compile functions, so it won't have any effect. I tried this in the first place before changing Lexy itself.

@aheinze What about this variant?

https://github.com/raffaelj/cockpit/commit/bfe3f6e7fedb6f3d38f3cb7e14951a3aede8a228

I added a kind of after event, so I can choose to call an extension at the end of the whole process.

After adding this to my bootstrap.php, my html is prettified.

```php
$app->renderer->extend(function($content) {
    return preg_replace('/([\r\n])(\s*)\<\?php/', '$1<?php', $content);
}, true);
```

**Comment by raffaelj** [comment]
I found a simple way to extend the Lexy parser without touching the core class. Now the html output of CpMultiplane is much more readable. I don't need the modification in the core anymore, so I'll close this issue. In case someone else wants to extend Lexy in a similar way, here is the code:

https://github.com/raffaelj/CpMultiplane/commit/ab856809636abe30844c2d4a45f1561f2ed6c92f

bootstrap.php:

```php
// use a modified renderer, that extends the core Lexy class
$lexy = $this->renderer;
$this->service('renderer', function() use ($lexy) {

    $renderer = new \MPLexy($lexy);

    // remove some white space to prettify the html output
    $renderer->after(function($content) {
        return preg_replace('/([\r\n])(\s*)\<\?php(?!\s*(echo|\$app->trigger))/', '$1<?php', $content);
    });

    return $renderer;
});
```

custom MPLexy class:

```php
<?php
/**
 * Extend Lexy template parser to call compilers after all extensions
 * and core compilers are done.
 *
 * @see https://github.com/agentejo/cockpit/blob/next/lib/Lexy.php
 */

class MPLexy extends \Lexy {

    protected $extensionsAfter = [];

    // construct with existing extensions from core Lexy
    public function __construct($lexy = null) {

        if ($lexy && property_exists($lexy, 'extensions')) {
            $this->extensions = $lexy->extensions;
        }

        // add compiler to the end of the list
        $this->compilers[] = 'after';

    }

    // like extend function
    public function after($compiler) {
        $this->extensionsAfter[] = $compiler;
    }

    // like compile_extensions function, but called at the end
    protected function compile_after($value) {

        foreach ($this->extensionsAfter as &$compiler) {
            $value = call_user_func($compiler, $value);
        }

        return $value;
    }

}
```
